sklearn/model_selection/_validation.py

```diff
@@ -727,12 +727,43 @@ def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params,
     predictions = func(X_test)
     if method in ['decision_function', 'predict_proba', 'predict_log_proba']:
         n_classes = len(set(y))
-        predictions_ = np.zeros((_num_samples(X_test), n_classes))
-        if method == 'decision_function' and len(estimator.classes_) == 2:
-            predictions_[:, estimator.classes_[-1]] = predictions
-        else:
-            predictions_[:, estimator.classes_] = predictions
-        predictions = predictions_
+        if n_classes != len(estimator.classes_):
+            if method == 'predict_proba':
+                # Fill missing classes with zero
+                predictions_ = np.zeros((_num_samples(X_test), n_classes))
+                predictions_[:, estimator.classes_] = predictions
+
+            elif method == 'predict_log_proba':
+                # Fill missing classes with minimum value
+                predictions_ = np.full((_num_samples(X_test), n_classes),
+                                       np.finfo(predictions.dtype).min)
+                predictions_[:, estimator.classes_] = predictions
+
+            else:  # Special handling logic for decision_function
+                err_mess = 'Output shape {} of {} does not match ' \
+                           'number of classes ({}) in fold. Cannot' \
+                           ' reconcile different number of classes' \
+                           ' in different folds. To fix this, use ' \
+                           'a cross-validation technique resulting' \
+                           ' in properly stratified folds'
+                if predictions.ndim == 2 and \
+                        predictions.shape[1] != len(estimator.classes_):
+                    # This handles the case when the shape of predictions
+                    # does not match the number of classes used to train
+                    # it with. This case is found when sklearn.svm.SVC is
+                    # set to `decision_function_shape='ovo'`.
+                    raise ValueError(err_mess.format(predictions.shape, method,
+                                                     len(estimator.classes_)))
+                if len(estimator.classes_) <= 2:
+                    # In this special case, `predictions` contains a 1D array.
+                    raise ValueError(err_mess.format(predictions.shape, method,
+                                                     len(estimator.classes_)))
+
+                predictions_ = np.full((_num_samples(X_test), n_classes),
+                                       np.finfo(predictions.dtype).min)
+                predictions_[:, estimator.classes_] = predictions
```

Comment:
What changes for each method is the default value, you could simplify the code and use better variable names, e.g.  something like this in the non-error case:

```py
float_min = np.finfo(predictions.dtype).min
default_values = {'decision_function': float_min, 'predict_log_proba': float_min, 
    'predict_proba': 0}
predictions_for_all_classes = np.full((_num_samples(predictions), n_classes),
    default_values[method])
predictions_for_all_classes[:, estimator.classes_] = predictions
```